File: FBSimulatorControl/Management/FBSimulatorSet.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorSet.h"
#import "FBSimulatorSet+Private.h"

#import <CoreSimulator/SimDevice.h>
#import <CoreSimulator/SimDeviceSet.h>
#import <CoreSimulator/SimDeviceType.h>
#import <CoreSimulator/SimRuntime.h>
#import <CoreSimulator/SimServiceContext.h>

#import <FBControlCore/FBControlCore.h>

#import <objc/runtime.h>

#import "FBCoreSimulatorNotifier.h"
#import "FBSimulatorControl.h"
#import "FBSimulatorControlConfiguration.h"
#import "FBSimulatorControlFrameworkLoader.h"
#import "FBSimulatorDeletionStrategy.h"
#import "FBSimulatorEraseStrategy.h"
#import "FBSimulatorInflationStrategy.h"
#import "FBSimulatorNotificationUpdateStrategy.h"
#import "FBSimulatorShutdownStrategy.h"

@implementation FBSimulatorSet

@synthesize allSimulators = _allSimulators;
@synthesize delegate = _delegate;

#pragma mark Initializers

+ (void)initialize
{
  [FBSimulatorControlFrameworkLoader.essentialFrameworks loadPrivateFrameworksOrAbort];
}

+ (instancetype)setWithConfiguration:(FBSimulatorControlConfiguration *)configuration deviceSet:(SimDeviceSet *)deviceSet delegate:(id<FBiOSTargetSetDelegate>)delegate logger:(id<FBControlCoreLogger>)logger reporter:(id<FBEventReporter>)reporter error:(NSError **)error
{
  return [[FBSimulatorSet alloc] initWithConfiguration:configuration deviceSet:deviceSet delegate:delegate logger:logger reporter:reporter];
}

- (instancetype)initWithConfiguration:(FBSimulatorControlConfiguration *)configuration deviceSet:(SimDeviceSet *)deviceSet delegate:(id<FBiOSTargetSetDelegate>)delegate logger:(id<FBControlCoreLogger>)logger reporter:(id<FBEventReporter>)reporter
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _configuration = configuration;
  _deviceSet = deviceSet;
  _delegate = delegate;
  _logger = logger;
  _reporter = reporter;
  _workQueue = configuration.workQueue ? configuration.workQueue : dispatch_get_main_queue();
  _asyncQueue = configuration.asyncQueue ? configuration.asyncQueue : dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);

  _allSimulators = @[];
  _inflationStrategy = [FBSimulatorInflationStrategy strategyForSet:self];
  _notificationUpdateStrategy = [FBSimulatorNotificationUpdateStrategy strategyWithSet:self];

  return self;
}

#pragma mark Querying

- (id<FBiOSTargetInfo>)targetWithUDID:(NSString *)udid
{
  return [self simulatorWithUDID:udid];
}

- (FBSimulator *)simulatorWithUDID:(NSString *)udid
{
  return [[self.allSimulators filteredArrayUsingPredicate:FBiOSTargetPredicateForUDID(udid)] firstObject];
}

#pragma mark Creation

- (FBFuture<FBSimulator *> *)createSimulatorWithConfiguration:(FBSimulatorConfiguration *)configuration
{
  FBDeviceModel model = configuration.device.model;

  // See if we meet the runtime requirements to create a Simulator with the given configuration.
  NSError *innerError = nil;
  SimDeviceType *deviceType = [configuration obtainDeviceTypeWithError:&innerError];
  if (!deviceType) {
    return [[[FBSimulatorError
      describeFormat:@"Could not obtain a DeviceType for Configuration %@", configuration]
      causedBy:innerError]
      failFuture];
  }
  SimRuntime *runtime = [configuration obtainRuntimeWithError:&innerError];
  if (!runtime) {
    return [[[FBSimulatorError
      describeFormat:@"Could not obtain a SimRuntime for Configuration %@", configuration]
      causedBy:innerError]
      failFuture];
  }

  // First, create the device.
  [self.logger.debug logFormat:@"Creating device with Type %@ Runtime %@", deviceType, runtime];
  return [[[FBSimulatorSet
    onDeviceSet:self.deviceSet createDeviceWithType:deviceType runtime:runtime name:model queue:self.asyncQueue]
    onQueue:self.workQueue fmap:^(SimDevice *device) {
      return [self fetchNewlyMadeSimulator:device];
    }]
    onQueue:self.workQueue fmap:^(FBSimulator *simulator) {
      simulator.configuration = configuration;
      [self.logger.debug logFormat:@"Created Simulator %@ for configuration %@", simulator.udid, configuration];

      // This step ensures that the Simulator is in a known-shutdown state after creation.
      // This prevents racing with any 'booting' interaction that occurs immediately after allocation.
      return [[[FBSimulatorShutdownStrategy
        shutdown:simulator]
        rephraseFailure:@"Could not get newly-created simulator into a shutdown state"]
        mapReplace:simulator];
    }];
}

- (FBFuture<FBSimulator *> *)cloneSimulator:(FBSimulator *)simulator toDeviceSet:(FBSimulatorSet *)destinationSet
{
  NSParameterAssert(simulator.set == self);
  return [[FBSimulatorSet
    onDeviceSet:self.deviceSet cloneDevice:simulator.device toDeviceSet:destinationSet.deviceSet queue:self.asyncQueue]
    onQueue:self.workQueue fmap:^(SimDevice *device) {
      return [destinationSet fetchNewlyMadeSimulator:device];
    }];
}

- (NSArray<FBSimulatorConfiguration *> *)configurationsForAbsentDefaultSimulators
{
  NSSet<FBSimulatorConfiguration *> *existingConfigurations = [NSSet setWithArray:[self.allSimulators valueForKey:@"configuration"]];
  NSMutableSet<FBSimulatorConfiguration *> *absentConfigurations = [NSMutableSet setWithArray:[FBSimulatorConfiguration allAvailableDefaultConfigurationsWithLogger:self.logger]];
  [absentConfigurations minusSet:existingConfigurations];
  return [absentConfigurations allObjects];
}

#pragma mark Destructive Methods

- (FBFuture<NSNull *> *)shutdown:(FBSimulator *)simulator
{
  NSParameterAssert(simulator);
  return [FBSimulatorShutdownStrategy shutdown:simulator];
}

- (FBFuture<NSNull *> *)erase:(FBSimulator *)simulator
{
  NSParameterAssert(simulator);
  return [FBSimulatorEraseStrategy erase:simulator];
}

- (FBFuture<NSNull *> *)delete:(FBSimulator *)simulator
{
  NSParameterAssert(simulator);
  return [FBSimulatorDeletionStrategy delete:simulator];
}

- (FBFuture<NSNull *> *)shutdownAll:(NSArray<FBSimulator *> *)simulators
{
  NSParameterAssert(simulators);
  return [FBSimulatorShutdownStrategy shutdownAll:simulators];
}

- (FBFuture<NSNull *> *)deleteAll:(NSArray<FBSimulator *> *)simulators;
{
  NSParameterAssert(simulators);
  return [FBSimulatorDeletionStrategy deleteAll:simulators];
}

- (FBFuture<NSNull *> *)shutdownAll
{
  NSArray<FBSimulator *> *simulators = self.allSimulators;
  return [FBSimulatorShutdownStrategy shutdownAll:simulators];
}

- (FBFuture<NSNull *> *)deleteAll
{
  return [self deleteAll:self.allSimulators];
}

#pragma mark NSObject

- (NSString *)description
{
  return [FBCollectionInformation oneLineDescriptionFromArray:self.allSimulators];
}

#pragma mark Private Methods

+ (NSDictionary<NSString *, FBSimulator *> *)keySimulatorsByUDID:(NSArray *)simulators
{
  NSMutableDictionary<NSString *, FBSimulator *> *dictionary = [NSMutableDictionary dictionary];
  for (FBSimulator *simulator in simulators) {
    dictionary[simulator.udid] = simulator;
  }
  return [dictionary copy];
}

- (FBFuture<FBSimulator *> *)fetchNewlyMadeSimulator:(SimDevice *)device
{
  // The SimDevice should now be in the DeviceSet and thus in the collection of Simulators.
  FBSimulator *simulator = [FBSimulatorSet keySimulatorsByUDID:self.allSimulators][device.UDID.UUIDString];
  if (!simulator) {
    return [[FBSimulatorError
      describeFormat:@"Expected simulator with UDID %@ to be inflated", device.UDID.UUIDString]
      failFuture];
  }
  return [FBFuture futureWithResult:simulator];
}

#pragma mark Public Properties

- (NSArray<FBSimulator *> *)allSimulators
{
  _allSimulators = [[self.inflationStrategy
    inflateFromDevices:self.deviceSet.availableDevices exitingSimulators:_allSimulators]
    sortedArrayUsingSelector:@selector(compare:)];
  return _allSimulators;
}

#pragma mark FBiOSTargetSet Implementation

- (NSArray<id<FBiOSTarget>> *)allTargetInfos
{
  return self.allSimulators;
}

#pragma mark Private Properties

+ (FBFuture<SimDevice *> *)onDeviceSet:(SimDeviceSet *)deviceSet createDeviceWithType:(SimDeviceType *)deviceType runtime:(SimRuntime *)runtime name:(NSString *)name queue:(dispatch_queue_t)queue
{
  FBMutableFuture<SimDevice *> *future = FBMutableFuture.future;
  [deviceSet createDeviceAsyncWithType:deviceType runtime:runtime name:name completionQueue:queue completionHandler:^(NSError *error, SimDevice *device) {
    if (device) {
      [future resolveWithResult:device];
    } else {
      [future resolveWithError:error];
    }
  }];
  return future;
}

+ (FBFuture<SimDevice *> *)onDeviceSet:(SimDeviceSet *)deviceSet cloneDevice:(SimDevice *)device toDeviceSet:(SimDeviceSet *)destinationSet queue:(dispatch_queue_t)queue
{
  FBMutableFuture<SimDevice *> *future = FBMutableFuture.future;
  [deviceSet cloneDeviceAsync:device name:device.name toSet:destinationSet completionQueue:queue completionHandler:^(NSError *error, SimDevice *created) {
    if (created) {
      [future resolveWithResult:created];
    } else {
      [future resolveWithError:error];
    }
  }];
  return future;
}

@end

File: FBSimulatorControl/Management/FBSimulatorSet.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@preconcurrency import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

public final class FBSimulatorSet: FBiOSTargetSet {

  // MARK: - Properties

  public let configuration: FBSimulatorControlConfiguration
  public let deviceSet: SimDeviceSet
  public weak var delegate: (any FBiOSTargetSetDelegate)?
  public let logger: any FBControlCoreLogger
  public let workQueue: DispatchQueue
  public let asyncQueue: DispatchQueue

  private var _allSimulators: [FBSimulator]
  // Guards _allSimulators: the allSimulators getter re-inflates and swaps the
  // backing array, and is called from arbitrary threads (lookups at companion
  // startup, delegate notification, description). Unsynchronized, the swap
  // races iteration in concurrent callers.
  private let simulatorsLock = NSLock()
  private lazy var inflationStrategy = FBSimulatorInflationStrategy.strategy(for: self)

  // Held only so that the strategy's notifier stays registered for the lifetime of the set; it is never read.
  private var notificationUpdateStrategy: FBSimulatorNotificationUpdateStrategy?

  // MARK: - Initializers

  /// - Parameter logger: nil means `FBControlCoreGlobalConfiguration.defaultLogger`, which is
  ///   os_log-only unless the `FBCONTROLCORE_LOGGING`/`FBCONTROLCORE_DEBUG_LOGGING` environment
  ///   variables are set — see its documentation. The resolved logger is stored non-optionally.
  public class func set(withConfiguration configuration: FBSimulatorControlConfiguration, deviceSet: SimDeviceSet, delegate: (any FBiOSTargetSetDelegate)?, logger: (any FBControlCoreLogger)?) throws -> FBSimulatorSet {
    let resolvedLogger = logger ?? FBControlCoreGlobalConfiguration.defaultLogger
    try FBSimulatorControlFrameworkLoader.essentialFrameworks.loadPrivateFrameworks(resolvedLogger)
    return FBSimulatorSet(configuration: configuration, deviceSet: deviceSet, delegate: delegate, logger: resolvedLogger)
  }

  private init(configuration: FBSimulatorControlConfiguration, deviceSet: SimDeviceSet, delegate: (any FBiOSTargetSetDelegate)?, logger: any FBControlCoreLogger) {
    self.configuration = configuration
    self.deviceSet = deviceSet
    self.delegate = delegate
    self.logger = logger
    self.workQueue = DispatchQueue.main
    self.asyncQueue = DispatchQueue.global(qos: .default)
    self._allSimulators = []
    self.notificationUpdateStrategy = FBSimulatorNotificationUpdateStrategy.strategy(with: self)
  }

  // MARK: - Querying

  public func target(withUDID udid: String) -> (any FBiOSTargetInfo)? {
    return simulator(withUDID: udid)
  }

  public func simulator(withUDID udid: String) -> FBSimulator? {
    return allSimulators.filter { FBiOSTargetPredicateForUDID(udid).evaluate(with: $0) }.first
  }

  // MARK: - Creation

  public func createSimulator(with configuration: FBSimulatorConfiguration) async throws -> FBSimulator {
    let model: String = configuration.device.model.rawValue

    let deviceType: SimDeviceType
    let runtime: SimRuntime
    do {
      deviceType = try configuration.obtainDeviceType()
      runtime = try configuration.obtainRuntime()
    } catch {
      throw FBSimulatorSetError.deviceTypeOrRuntimeUnavailable(configuration: "\(configuration)", reason: error.localizedDescription)
    }

    logger.debug().log("Creating device with Type \(deviceType) Runtime \(runtime)")
    let device = try await Self.createDevice(on: deviceSet, type: deviceType, runtime: runtime, name: model, queue: asyncQueue)
    let simulator = try fetchNewlyMadeSimulatorOrThrow(device)
    simulator.configuration = configuration
    logger.debug().log("Created Simulator \(simulator.udid) for configuration \(configuration)")
    do {
      try await FBSimulatorShutdownStrategy.shutdown(simulator)
    } catch {
      throw FBSimulatorSetError.shutdownAfterCreateFailed(reason: error.localizedDescription)
    }
    return simulator
  }

  public func cloneSimulator(_ simulator: FBSimulator, toDeviceSet destinationSet: FBSimulatorSet) async throws -> FBSimulator {
    let device = try await Self.cloneDevice(on: deviceSet, device: simulator.device, toDeviceSet: destinationSet.deviceSet, queue: asyncQueue)
    return try destinationSet.fetchNewlyMadeSimulatorOrThrow(device)
  }

  func configurationsForAbsentDefaultSimulators() throws -> [FBSimulatorConfiguration] {
    let existingConfigurations = Set(allSimulators.compactMap { $0.configuration })
    var absentConfigurations = Set(try FBSimulatorConfiguration.allAvailableDefaultConfigrations(withLogger: logger))
    absentConfigurations.subtract(existingConfigurations)
    return Array(absentConfigurations)
  }

  // MARK: - Destructive Methods

  public func shutdown(_ simulator: FBSimulator) async throws {
    try await FBSimulatorShutdownStrategy.shutdown(simulator)
  }

  public func delete(_ simulator: FBSimulator) async throws {
    try await FBSimulatorDeletionStrategy.delete(simulator)
  }

  func shutdownAll(_ simulators: [FBSimulator]) async throws {
    try await FBSimulatorShutdownStrategy.shutdownAll(simulators)
  }

  public func deleteAll(_ simulators: [FBSimulator]) async throws {
    try await FBSimulatorDeletionStrategy.deleteAll(simulators)
  }

  func shutdownAll() async throws {
    try await FBSimulatorShutdownStrategy.shutdownAll(allSimulators)
  }

  public func deleteAll() async throws {
    try await deleteAll(allSimulators)
  }

  // MARK: - Description

  public var description: String {
    FBCollectionInformation.oneLineDescription(from: allSimulators)
  }

  // MARK: - FBiOSTargetSet

  public var allTargetInfos: [any FBiOSTargetInfo] {
    allSimulators
  }

  // MARK: - Public Properties

  public var allSimulators: [FBSimulator] {
    simulatorsLock.lock()
    defer { simulatorsLock.unlock() }
    _allSimulators = inflationStrategy.inflate(
      fromDevices: deviceSet.availableDevices,
      exitingSimulators: _allSimulators
    )
    .sorted { ($0 as FBSimulator).compare($1 as any FBiOSTarget) == .orderedAscending }
    return _allSimulators
  }

  // MARK: - Private Methods

  private class func keySimulatorsByUDID(_ simulators: [FBSimulator]) -> [String: FBSimulator] {
    var dictionary: [String: FBSimulator] = [:]
    for simulator in simulators {
      dictionary[simulator.udid] = simulator
    }
    return dictionary
  }

  private func fetchNewlyMadeSimulatorOrThrow(_ device: SimDevice) throws -> FBSimulator {
    guard let simulator = FBSimulatorSet.keySimulatorsByUDID(allSimulators)[device.udid.uuidString] else {
      throw FBSimulatorSetError.simulatorNotInflated(udid: device.udid.uuidString)
    }
    return simulator
  }

  private static func createDevice(on deviceSet: SimDeviceSet, type deviceType: SimDeviceType, runtime: SimRuntime, name: String, queue: DispatchQueue) async throws -> SimDevice {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<SimDevice, Error>) in
      deviceSet.createDeviceAsync(withType: deviceType, runtime: runtime, name: name, completionQueue: queue) { error, device in
        if let device {
          continuation.resume(returning: device)
        } else {
          continuation.resume(throwing: error ?? FBSimulatorSetError.deviceCreationFailed)
        }
      }
    }
  }

  private static func cloneDevice(on deviceSet: SimDeviceSet, device: SimDevice, toDeviceSet destinationSet: SimDeviceSet, queue: DispatchQueue) async throws -> SimDevice {
    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<SimDevice, Error>) in
      deviceSet.cloneDeviceAsync(device, name: device.name, to: destinationSet, completionQueue: queue) { error, created in
        if let created {
          continuation.resume(returning: created)
        } else {
          continuation.resume(throwing: error ?? FBSimulatorSetError.deviceCloneFailed)
        }
      }
    }
  }
}
